sdk/src/wallet/core/builder.rs

```diff
@@ -253,74 +255,80 @@ where
             storage_manager: tokio::sync::RwLock::new(storage_manager),
         });
 
-        todo!("single account");
-        // let mut accounts: Vec<Account<S>> = try_join_all(
-        //     accounts
-        //         .into_iter()
-        //         .map(|a| Account::new(a, wallet_inner.clone()).boxed()),
-        // )
-        // .await?;
+        let address = create_wallet_address(&*wallet_inner.secret_manager, bip_path.coin_type, bip_path.account).await?;
+
+        let bech32_hrp = wallet_inner.client.get_bech32_hrp().await?;
+        // TODO: just take the last 10 chars or so
+        let alias = self
+            .alias
+            .unwrap_or_else(|| format!("{}", address.to_bech32_unchecked(bech32_hrp)));
+
+        let wallet_data = Arc::new(RwLock::new(WalletData::new(bip_path, address, bech32_hrp, alias)));
+
+        let wallet = Wallet {
+            inner: wallet_inner,
+            data: wallet_data,
+        };
 
         // If the wallet builder is not set, it means the user provided it and we need to update the addresses.
         // In the other case it was loaded from the database and addresses are up to date.
         if new_provided_client_options {
-            todo!("single account");
-            // for account in accounts.iter_mut() {
-            //     // Safe to unwrap because we create the client if accounts aren't empty
-            //     account.update_account_bech32_hrp().await?;
-            // }
+            wallet.update_bech32_hrp().await?;
         }
 
-        todo!("remove unwraps");
-        let wallet_data = Arc::new(RwLock::new(WalletData::new(
-            self.alias.unwrap(),
-            self.bip_path.unwrap(),
-            self.address.unwrap(),
-        )));
-
-        Ok(Wallet {
-            inner: wallet_inner,
-            data: wallet_data,
-            default_sync_options: todo!("SyncOptions::default()"),
-            last_synced: todo!("Mutex::new(0)"),
-        })
+        Ok(wallet)
     }
 
     #[cfg(feature = "storage")]
     pub(crate) async fn from_wallet(wallet: &Wallet<S>) -> Self {
         Self {
-            alias: Some(wallet.data.read().await.alias.clone()),
-            bip_path: Some(wallet.data.read().await.bip_path.clone()),
-            address: Some(wallet.data.read().await.address.clone()),
+            bip_path: Some(wallet.data().await.bip_path.clone()),
+            address: Some(wallet.data().await.address.clone()),
+            bech32_hrp: Some(wallet.data().await.bech32_hrp.clone()),
+            alias: Some(wallet.data().await.alias.clone()),
             client_options: Some(wallet.client_options().await),
             storage_options: Some(wallet.storage_options.clone()),
             secret_manager: Some(wallet.secret_manager.clone()),
         }
     }
 }
 
+/// Generate the wallet address.
+pub(crate) async fn create_wallet_address<S: SecretManage>(
```

Comment:
Address should be just generated with the secret manager and provided to the wallet builder

Reply:
since the secret manager is provided to the builder as well, this could be also done internally, no?

Reply:
Maybe it shouldn't be provided to it anymore, not sure
But in either case we want the possibility to have a wallet with an NFT or account address which we can't generate with a secret manager, so there needs to be a way for the user to provide the address